### PR TITLE
♻️ Condense mcdoc runtime errors in a unified way

### PR DESCRIPTION
**Pull request by NeunEinser**
This should fix all bugs with error condensing and valid filtering definitions.

### CODE REVIEW

File: packages/mcdoc/src/runtime/checker/error.ts [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+	definition: CheckerTreeDefinitionNode<T>
+	errors: McdocRuntimeError<T>[]
+}
+
+export function condenseErrorsAndFilterSiblings<T>(
+	definitions: {
+		definition: CheckerTreeDefinitionNode<T>
+		errors: McdocRuntimeError<T>[]
+	}[],
+	context: McdocCheckerContext<T>,
```
Comment (misode) [inline packages/mcdoc/src/runtime/checker/error.ts:114]: Context is never used throughout this whole function and `condense`, but I assume you're keeping it in potentially for future use cases, and maybe logging?

File: packages/mcdoc/src/runtime/checker/error.ts [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+export interface ErrorCondensingDefinition<T> {
+	definition: CheckerTreeDefinitionNode<T>
+	errors: McdocRuntimeError<T>[]
+}
+
+export function condenseErrorsAndFilterSiblings<T>(
+	definitions: {
+		definition: CheckerTreeDefinitionNode<T>
+		errors: McdocRuntimeError<T>[]
+	}[],
```
Comment (misode) [inline packages/mcdoc/src/runtime/checker/error.ts:113]: This should be `definitions: ErrorCondensingDefinition<T>[]`

File: packages/mcdoc/src/runtime/checker/error.ts (hunk 1 above)
Comment (NeunEinser) [reply packages/mcdoc/src/runtime/checker/error.ts:114]: Ah, this was needed for `isAssignable` before I replaced it with `equals`

File: packages/mcdoc/src/runtime/checker/error.ts (hunk 2 above)
Comment (NeunEinser) [reply packages/mcdoc/src/runtime/checker/error.ts:113]: Yeah I think this was nuked by a merge conflict

File: packages/mcdoc/src/runtime/checker/error.ts [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+		condensedErrors: errors,
+	}
+}
+
+function condense<T, E extends McdocRuntimeError<T>>(
+	validDefinitions: ErrorCondensingDefinition<T>[],
+	context: McdocCheckerContext<T>,
+	is: (e: McdocRuntimeError<T>) => e is E,
+	equals: (a: E, b: E) => boolean,
+	combineAlternatives?: (errors: E[]) => E,
```
[the remark this answers is not in the document]
Comment (NeunEinser) [reply packages/mcdoc/src/runtime/checker/error.ts:224]: Isn't this the same (plus two curly braces)?

File: packages/mcdoc/src/runtime/checker/error.ts [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+			continue
+		}
+		definitions.push(
+			...group
+				.filter(v =>
+					!definitionGroups.some((og, oi) =>
+						(oi > i || og.length === 1)
+							&& predicate
+							? og.some(ov => predicate(v, ov))
+							: og.includes(v)
```
[the remark this answers is not in the document]
Comment (NeunEinser) [reply packages/mcdoc/src/runtime/checker/error.ts:345]: This actually should be improved a bit over what we had before, as the old condensing needed to do similar things, but just did a flatMap and had a "truer" O(n^2), whereas here it's pretty optimized and only checks things it actually needs to, and only when conflicting errors actually exist (if it's length one it takes the short circuit)

If we run into issues here, we probably have to make some sacrifices and accept some errors to contain duplicated messages, but I think this one should be pretty fine.

File: packages/mcdoc/src/runtime/checker/error.ts (hunk 2 above)
Comment (NeunEinser) [reply packages/mcdoc/src/runtime/checker/error.ts:113]: Ah yeah good catch, this broke when I copied stuff over again after merging `internal` error. :upside_down_face:

File: packages/mcdoc/src/runtime/checker/error.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (misode) [reply packages/mcdoc/src/runtime/checker/error.ts]: The lambda functions are named when you use this function

**Comment by misode** [comment]
Maybe it wouldn't hurt adding some tests for the more complicated examples where this PR improves the error condensing?

**Comment by NeunEinser** [comment]
> Maybe it wouldn't hurt adding some tests for the more complicated examples where this PR improves the error condensing?

Yeah I agree, I think I would like to do them separetely though before I need to fix more merge conflicts :D

Not sure I will have time before the weekend. ;)

File: packages/mcdoc/src/runtime/checker/index.ts [hunk 5]
```diff
@@ -1,7 +1,8 @@
-import * as core from '@spyglassmc/core'
-import { arrayToMessage, localeQuote, localize } from '@spyglassmc/locales'
+import { Range, Source } from '@spyglassmc/core'
+import type { CheckerContext, FullResourceLocation } from '@spyglassmc/core'
```
Comment (misode) [inline packages/mcdoc/src/runtime/checker/index.ts:2]: I prefer having the `core` prefix for these, as is this case with most other files

Comment (NeunEinser) [reply packages/mcdoc/src/runtime/checker/index.ts:2]: ah well, I usually just do that when there are lots of imports from one package, but I can change it later ¯\\\_(ツ)\_/¯

**Review by misode — APPROVED** [review]

**Pull request merged**